[#58520] add an endpoint to fetch the hierarchy

Kharonus opened the pull request:
# Ticket
[OP#58520](https://community.openproject.org/work_packages/58520)

# What are you trying to accomplish?
- we need an API endpoint to fetch the hierarchy structure of a custom field of type hierarchy
- the tree is represented as a depth-first flat list

# What approach did you choose and why?
- adding an endpoint GET /custom_fields/:id/items
- adding an endpoint GET /custom_field_items/:id (needed for the self links)
- adding API specification for both
- adding parameter validation contract for the tree structure
- amended legacy tests

## Hint for review

This review is easily separable into 

a) the API specification files, located under `/docs`
b) the code, in `/app`, `/lib/api`, and `/spec`

those could be separate PRs, but sadly I have those changes on the same commits.

Kharonus committed:
[#58520] add an endpoint to fetch the hierarchy

- https://community.openproject.org/work_packages/58520
- add API specification about the endpoint

brunopagno commented on docs/api/apiv3/components/examples/hierarchy_item_collection_filtered_response.yml:
```diff
@@ -0,0 +1,53 @@
+        id: 1338
+        label: Stormtroopers
+        short: ST
+        _links:
+          self:
+            href: /api/v3/custom_field/42/items/1338
+          parent:
+            href: /api/v3/custom_field/42/items/1337
+          children:
+            - href: /api/v3/custom_field/42/items/1340
+              name: ST-377
+            - href: /api/v3/custom_field/42/items/1341
+              name: ST-422
```
I think you're right. Looks simple and elegant.

![image](https://github.com/user-attachments/assets/7bca08ab-310f-4b38-a91e-b2ba9d1f184d)

brunopagno commented on docs/api/apiv3/components/examples/hierarchy_item_collection_response.yml:
```diff
@@ -0,0 +1,45 @@
+    elements:
+      - _type: HierarchyItem
+        id: 1337
+        label: null
+        short: null
```
This makes me think again if we should copy data to the root item, instead of keeping it only in the custom field.

I imagine that in most cases whoever is consuming the API will also have to fetch the custom field anyways, so this is probably a non issue.

Commenting just not to leave unsaid.

Kharonus replied:
I also think, the custom field is everytime loaded first. Without fetching it first, you won't have the id and also not the field format. Only with the format of `hierarchy` (and later `list`) the client is interested in retrieving any items for it.

brunopagno replied:
Yes. The only argument against would if if we ever want to extract hierarchy away from custom fields, but that would be solving a problem that does not exist yet.

So, all good :+1:

brunopagno reviewed (commented):
I say the API spec looks beautiful. A work of art. Deserving of praise.
:tada:

Kharonus committed:
[#58520] Implement fetching tree structure

- implement hashed subtree retrieval in persistence layer
- implement flattening of hashed subtree in applicaiton layer
- added api spec for GET /custom_field_items/:id route
- added self link, parent and children to http response
- added api for single hierarchy item route
- write unit tests for hashed_subtree in HierarchicalItemService
- add parameter validation to api endpoint

apfohl commented on lib/api/v3/custom_fields/hierarchy/items_api.rb:
```diff
@@ -0,0 +1,118 @@
+              sub_tree = ::CustomFields::Hierarchy::HierarchicalItemService
+                           .new
+                           .hashed_subtree(item: start_item, depth:)
+                           .value!
```
🟡 Should we handle an error case here?

Kharonus replied:
added an either handling

mereghost reviewed (commented):
LGTM.

Kharonus committed:
Merge branch 'dev' into implementation/58520-api-add-endpoint-for-fetching-the-tree
